LFS-743: As an API consumer, I can access the human-readable answers in addition to the answer value in the form JSON

veronikaslc opened the pull request:
[LFS-743](https://phenotips.atlassian.net/browse/LFS-423)

Commit:
LFS-743: As an API consumer, I can access the human-readable answers in addition to the answer value in the form JSON

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/FileLabelProcessor.java:
```diff
@@ -0,0 +1,120 @@
+    {
+        try {
+            List<String> names = new ArrayList<>();
+            NodeIterator childNodes = node.getNodes("*.csv");
```
Why "*.csv"?

marta- replied:
Why "*.csv"? We can have any type of files.

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/FileLabelProcessor.java:
```diff
@@ -0,0 +1,120 @@
+                    names.add(childNode.getName());
+                }
+            }
+            return StringUtils.join(names, ", ");
```
[Per specs](https://phenotips.atlassian.net/browse/LFS-743), if the `value` is an array, then `displayedValue` should also be an array:

> For questions with multiple answers, the displayedValue should be an array with 1-to-1 correspondence to the value array.

(this applies to all multi-valued answers, not just files)

The reason behind this is so that we have flexibility in displaying multiple values. On the patient chart we may indeed want to join them by `, `, whereas on the full read-only form we will probably display them one per line as a list.

Moreover, for files we will have to display download links, which means we need each `displayedValue` (file name) separately, to combine it with its corresponding `value` (file path) and generate the link on the frontend.

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/AnswerOptionsLabelProcessor.java:
```diff
@@ -0,0 +1,152 @@
+                propsSet.add(nodeProp.getString());
+            }
+
+            if (question != null) {
```
This can probably be turned into 
```
if (question == null) {
  // return propSet
}
```
to help with `checkstyle:NestedIfDepth`.

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/AnswerOptionsLabelProcessor.java:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    @Override
+    @SuppressWarnings({"checkstyle:CyclomaticComplexity", "checkstyle:NestedIfDepth"})
```
Please reorganize or split the function so that `@SuppressWarnings` won't be necessary.

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/BooleanLabelProcessor.java:
```diff
@@ -0,0 +1,103 @@
+    {
+        try {
+            int rawValue = (int) node.getProperty(PROP_VALUE).getLong();
+            return BooleanUtils.toString(BooleanUtils.toBooleanObject(rawValue, 1, 0, -1), "Yes", "No", "Unknown");
```
The labels for Boolean question options are configurable in the question definition, with "Yes", "No", "Unknown" being the defaults. See how they're handled here: https://github.com/ccmbioinfo/lfs/blob/dev/modules/data-entry/src/main/frontend/src/questionnaire/Subject.jsx#L524

sashaandjic commented:
Shouldn't the `displayValue `for `file `be:
`0: "Vmax.txt"` 
`1: "test 1.txt"`
![image](https://user-images.githubusercontent.com/23243244/102962228-612be200-44b4-11eb-8603-f86b6ea41b91.png)

sashaandjic commented:
Free text is not included in displayedValue:
![image](https://user-images.githubusercontent.com/23243244/102962328-989a8e80-44b4-11eb-92e2-29ed39be4eb5.png)

![image](https://user-images.githubusercontent.com/23243244/102962286-87518200-44b4-11eb-9251-c763aa9bca83.png)

Commit:
LFS-743: As an API consumer, I can access the human-readable answers in addition to the answer value in the form JSON
Addressing codereview comments

marta- reviewed (changes requested):
2 issues identified while testing this branch in `cardiac_rehab` mode using the "External file" questionnaire (see output of `.bare.deep.-identify.json` export below):

* For multi-valued questions, the entryes in the `displayedValue` array and those in the `value` array should be in the same order. For the `file` question in the export example below these are swapped, making it more difficult to generate file links on the frontend.
* For questions that accept multiple answers but only received one, the `value` isn't an array, while `displayedValue` is an array of 1 (see the `source` question in the example export below). They should both have the same data type. I suggest that `displayedValue` be turned into a string if there's only one answer for backwards compatibility, but I'm open to alternatives (e.g. making `value` an array instead) if they come with good justification.

```
{
	"subject": "we12345",
	"questionnaire": "External file",
	"created": "2020-12-23T19:12:26.538-05:00",
	"file": {
		"question": "File (de-identified)",
		"value": ["/Forms/ce8fdd9f-3787-4ba0-9e55-18270075798a/2802b6e2-47b4-4a60-8d25-f687eaaec694/LFS-650.png", "/Forms/ce8fdd9f-3787-4ba0-9e55-18270075798a/2802b6e2-47b4-4a60-8d25-f687eaaec694/file2.csv"],
		"file2.csv": {
			"content": " \n"
		},
		"LFS-650.png": {
			"content": "\u0089PNG\r\n\u001a\n"
		},
		"displayedValue": ["file2.csv", "LFS-650.png"]
	},
	"file-visit": {
		"question": "Visit number",
		"value": 7,
		"displayedValue": "7"
	},
	"source": {
		"question": "File source",
		"value": "Vmax",
		"displayedValue": ["Vmax"]
	}
}
```

Commit:
LFS-743: As an API consumer, I can access the human-readable answers in addition to the answer value in the form JSON
Retain the order ot the listed properties

sashaandjic commented:
The order in **value** and **display value** arrays is not the same. Also, should content also be presented in an array?
![image](https://user-images.githubusercontent.com/23243244/103375970-705cf200-4aa9-11eb-8c2e-ae4c8540adcc.png)

marta- reviewed (changes requested):
File answer order issue confirmed. The latest fix does not cover file questions.

sdumitriu commented:
The content ile should NOT be processed in any way. It is assumed to be opaque binary content.

Commit:
LFS-743: As an API consumer, I can access the human-readable answers in addition to the answer value in the form JSON
Retain the order of the listed file names

marta- commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/FileLabelProcessor.java:
```diff
@@ -0,0 +1,119 @@
+    {
+        try {
+            List<String> names = new ArrayList<>();
+            NodeIterator childNodes = node.getNodes("*.*");
```
Files without an extension are ignored and don't show up in the `displayedValue` list.

veronikaslc replied:
Could they be uploaded? Afair we only accept files in special format.

marta- replied:
Yes, they can be uploaded. For example, in cardiac_rehab there are no file name or extension restrictions in the "External file" form. In general we should be able to support any valid file.

marta- reviewed (changes requested):
The order of file names in `displayedValue` still differs from the order in `value`. I can't see ATM how using only the child nodes of type `file` can guarantee the same order as in `value`. Can we extract the file names from the file paths in `value` instead?

Commit:
LFS-743: As an API consumer, I can access the human-readable answers in addition to the answer value in the form JSON
Fixed files order

Commit:
LFS-743: As an API consumer, I can access the human-readable answers in addition to the answer value in the form JSON
displayedValue should be turned into a string if there's only one answer

sashaandjic reviewed (approved):
tested in LFS and cardiac_rehab mode - all good

marta- reviewed (approved):
Looks good to me

sdumitriu committed:
LFS-743: As an API consumer, I can access the human-readable answers in addition to the answer value in the form JSON
Moved all the label processors in a separate package

sdumitriu committed:
LFS-743: As an API consumer, I can access the human-readable answers in addition to the answer value in the form JSON
Move all common methods in the parent class
Make the common parent class abstract

sdumitriu committed:
LFS-743: As an API consumer, I can access the human-readable answers in addition to the answer value in the form JSON
Bugfix: date processor doesn't support multi-valued properties

sdumitriu committed:
[cleanup] Applied codestyle

sdumitriu committed:
LFS-743: As an API consumer, I can access the human-readable answers in addition to the answer value in the form JSON
Reduced visibility of utility methods

sdumitriu committed:
LFS-743: As an API consumer, I can access the human-readable answers in addition to the answer value in the form JSON
Bugfix: Default label processor does not support multivalued properties

sdumitriu committed:
LFS-743: As an API consumer, I can access the human-readable answers in addition to the answer value in the form JSON
Make the default label processor work for all answer types for better extensibility

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/AnswerOptionsLabelProcessor.java:
```diff
@@ -0,0 +1,157 @@
+    public boolean canProcess(Resource resource)
+    {
+        return resource.isResourceType("lfs/Form");
+    }
```
These are all the same in every processor, so they should be moved to the common `SimpleAnswer`

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/AnswerOptionsLabelProcessor.java:
```diff
@@ -0,0 +1,157 @@
+            // Really shouldn't happen
+        }
+        return input;
+    }
```
This isn't needed, the `deep` processor does this for every child, regardless of type. If the goal is to instead make sure that (some) answers are always included in the JSON, even if `.deep` is not specified, I don't think that is needed.

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/AnswerOptionsLabelProcessor.java:
```diff
@@ -0,0 +1,157 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package ca.sickkids.ccm.lfs.dataentry.internal.serialize;
```
There are a lot of processors for the same target, so they deserve their own subpackage.

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/DateLabelProcessor.java:
```diff
@@ -0,0 +1,110 @@
+    {
+        try {
+            if (question != null) {
+                Calendar rawValue = node.getProperty(PROP_VALUE).getDate();
```
While in most cases a date question will have only one value, there is nothing that would prevent it from having more than one. In that case, this will fail with a `ValueFormatException`, leading to a `NullPointerException` when trying to add a `null` in the JSON, which will stop the whole JSON generation. It would be good to have all the label processors handle the multivalue case, even if not expected to have multivalued answers right now.

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/DefaultLabelProcessor.java:
```diff
@@ -0,0 +1,96 @@
+public class DefaultLabelProcessor extends SimpleAnswerLabelProcessor implements ResourceJsonProcessor
+{
+    private static final List<String> DEFAULT_RESOURCE_TYPES =
+            Arrays.asList("lfs:LongAnswer", "lfs:DoubleAnswer", "lfs:DecimalAnswer", "lfs:TimeAnswer",
```
This is not extensible, every new answer type will have to be added here. A better approach is for this to handle all `lfs:Answer` nodes, with each other custom processor overriding the default. This works because:

- this can have a lower priority so that it runs before all other custom processors
- `JsonObjectBuilder.add` overrides previously set values
- we don't expect any answer that would have a `value`, but not a `displayedValue`

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/FileLabelProcessor.java:
```diff
@@ -0,0 +1,129 @@
+    public void leave(Node node, JsonObjectBuilder json, Function<Node, JsonValue> serializeNode)
+    {
+        try {
+            if (node.isNodeType("lfs:FileResourceAnswer") || node.isNodeType("lfs:SomaticVariantsAnswer")) {
```
It's not proper to have a reference to `SomaticVariantsAnswer` in here, since that type is defined in a different module. It would be better to copy this processor into the other module, with each copy handling only one type.

sdumitriu commented on modules/data-entry/src/main/java/ca/sickkids/ccm/lfs/dataentry/internal/serialize/SimpleAnswerLabelProcessor.java:
```diff
@@ -0,0 +1,126 @@
+    public JsonValue getAnswerLabel(final Node node, final Node question)
+    {
+        try {
+            String value = node.getProperty(PROP_VALUE).getValue().toString();
```
This one also doesn't handle multivalued properties.

Commit:
LFS-743: As an API consumer, I can access the human-readable answers in addition to the answer value in the form JSON
File label processor should handle single value case

Commit:
LFS-743: As an API consumer, I can access the human-readable answers in addition to the answer value in the form JSON
Code optimisation

sashaandjic commented:
retested
